pxtpy/converter.ts

```diff
@@ -1307,23 +1338,71 @@ namespace pxt.py {
                     nodes.push(B.mkCommaSep(n.bases.map(expr)))
                     let b = getClassDef(n.bases[0])
                     if (b) {
-                        n.baseClass = b
+                        n.baseClass = b.symInfo
                         sym.extendsTypes = [b.symInfo.pyQName!]
                     }
+                    else {
+                        const nm = tryGetName(n.bases[0]);
+
+                        if (nm) {
+                            const localSym = lookupSymbol(nm);
+                            const globalSym = lookupGlobalSymbol(nm);
+
+                            n.baseClass = localSym || globalSym;
+
+                            if (n.baseClass) sym.extendsTypes = [n.baseClass.pyQName!]
+                        }
+                    }
                 }
             }
-            let body = stmts(n.body)
+
+            const classDefs = n.body.filter(s => n.isNamespace || s.kind === "FunctionDef");
+            const staticStmts = n.isNamespace ? [] : n.body.filter(s => classDefs.indexOf(s) === -1);
+
+            let body = stmts(classDefs)
             nodes.push(body)
 
-            let fieldDefs = listClassFields(n)
-                .filter(f => f.kind == SK.Property && f.isInstance)
-                .map(f => {
-                    if (!f.pyName || !f.pyRetType)
-                        error(n, 9535, lf("field definition missing py name or ret type", f.qName));
-                    return f
-                })
-                .map((f) => B.mkStmt(accessAnnot(f), quote(f.pyName!), typeAnnot(f.pyRetType!)))
-            body.children = fieldDefs.concat(body.children)
+            // Python classes allow arbitrary statements in their bodies, sort of like namespaces.
+            // Take all of these statements and put them in a static method that we can call when
+            // the class is defined.
+            let generatedInitFunction = false;
+            if (staticStmts.length) {
+                generatedInitFunction = true;
+
+                const staticBody = stmts(staticStmts);
+
+                const initFun = B.mkStmt(B.mkGroup([
+                    B.mkText(`public static __init${n.name}() `),
+                    staticBody
+                ]))
+
+                body.children.unshift(initFun);
+            }
+
+            if (!n.isNamespace) {
+                let isStatic = (f: SymbolInfo) => f.kind === SK.Property && !f.isInstance || f.kind === SK.Variable;
+
+                const fieldDefs = listClassFields(n, false)
+                    .filter(f => f.kind == SK.Property || isStatic(f))
+                    .map(f => {
+                        if (!f.pyName || !f.pyRetType)
+                            error(n, 9535, lf("field definition missing py name or ret type", f.qName));
```

Comment:
```suggestion
                            error(n, 9535, lf("field definition missing py name or return type", f.qName));
```

sidenote: hm I suppose that user would probably be able to infer what py name refers to by looking at the line? We should maybe go through all the errors on this page and put up a python-errors docs page that just gives a little more explanation and an example of fixing. not directly relevant to this pr tho